### DISCUSSION TITLE
All samples in a given batch are not returned when using DataListLoader and DataParallel

### DISCUSSION THREAD

**Discussion by tommysisk** [pyg-team/pytorch_geometric#6214] [Q&A] [answered]
I am attempting to use DataParallel and DataListLoader to train the graph neural network shown below (ignoring all but the forward method) on 8 GPUs. The network takes a Heterograph as input and returns a softmax output. When I run a single sample through the network without applying DataParallel, I get the expected softmax output. However, when I apply DataParallel to the network module and pass a batch/list of graphs from DataListLoader through it in a training loop, I am only returned 8 tensors. Upon further inspection, I found that the number of outputs I get == the number of GPUs I specify in the DataParallel call. Below the forward method of the network I show the output I get outside of the training loop without applying DataParallel and after that I show the result of passing data through the model in a training loop where DataParallel is applied and the batch is loaded from DataListLoader.  Simply put, I want to return the softmax output for all samples in the batch with the correct grad_fn (softmax) so I can input them into a user defined loss function i.e. a loss function not offered by PYG or regular torch. I am new to graph neural networks and am likely at fault here but I can't see any obvious reason why this problem would arise based on docs and example. Please let me know what I'm doing wrong :)

<img width="880" alt="Screen Shot 2022-12-12 at 4 06 01 AM" src="https://user-images.githubusercontent.com/85724985/207005202-f1a83391-3d39-4cfa-b989-3ee0c3a8ad46.png">
<img width="997" alt="Screen Shot 2022-12-12 at 3 55 45 AM" src="https://user-images.githubusercontent.com/85724985/207002908-111ac319-dd16-4076-84d7-ea2ec5c3be8b.png">
<img width="827" alt="Screen Shot 2022-12-12 at 3 58 37 AM" src="https://user-images.githubusercontent.com/85724985/207003568-0fa6f270-d1fe-4add-bc09-4397a6a25179.png">
<img width="977" alt="Screen Shot 2022-12-12 at 4 05 22 AM" src="https://user-images.githubusercontent.com/85724985/207005063-95e015fe-70ad-4012-a407-fce1494ea49b.png">

**Comment by rusty1s** [comment, marked as answer]
I am not totally sure if `DataListLoader` works with `HeteroData` objects TBH. The recommended approach is to use `DistributedDataParallel` for this. Here is an example: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/multi_gpu/distributed_batching.py which should work seamlessly with `HeteroData` objects as well.

**Reply by tommysisk** [reply]
Indeed, seamless! Thank you!!
